audb/core/repository.py

```diff
@@ -1,6 +1,90 @@
+import typing
+
 import audbackend
 
 
-class Repository(audbackend.Repository):
-    __doc__ = audbackend.Repository.__doc__
-    pass
+class Repository:
+    r"""Repository object.
+
+    It stores all information
+    needed to address a repository:
+    the repository name,
+    host,
+    and the backend name.
+
+    Args:
+        name: repository name
+        host: repository host
+        backend: repository backend
+
+    Examples:
+        >>> Repository("data-local", "/data", "file-system")
+        Repository('data-local', '/data', 'file-system')
+
+    """
+
+    backend_registry = {
+        "file-system": audbackend.backend.FileSystem,
+        "artifactory": audbackend.backend.Artifactory,
+    }
+    r"""Backend registry."""
+
+    def __init__(
+        self,
+        name: str,
+        host: str,
+        backend: str,
+    ):
+        self.name = name
+        r"""Repository name."""
+        self.host = host
+        r"""Repository host."""
+        self.backend = backend
+        r"""Repository backend."""
+
+    def __repr__(self):  # noqa: D105
+        return (
+            f"Repository("
+            f"'{self.name}', "
+            f"'{self.host}', "
+            f"'{self.backend}'"
+            f")"
+        )
+
+    def __call__(self) -> typing.Type[audbackend.interface.Base]:
+        r"""Return interface to access repository.
+
+        Returns:
+            interface to repository
+
+        """
+        backend_class = self.backend_registry[self.backend]
+        backend = backend_class(self.host, self.name)
+        if self.backend == "artifactory":
+            # Maven version file structure on Artifactory
+            interface = audbackend.interface.Maven(backend)
+        else:
+            interface = audbackend.interface.Versioned(backend)
+        return interface
+
+    @classmethod
+    def register(
+        cls,
+        backend_name: str,
+        backend_class: typing.Type[audbackend.backend.Base],
```

Comment:
This is an arbitrary place for putting this comment, but:
I find that audbackend has quite a few abstract data types, and I wonder whether it will start make sense to collect these into a separate `.pyi` file? I mean typing complexity in this package has again increased with the addition of the  `interface` type.
I think that .pyi generation can even be automated.

It of course might still be that this is still over the top.

Reply:
Interesting, I wasn't aware of `.pyi` files. I also have not completely understood them. https://mypy.readthedocs.io/en/stable/stubs.html seems to indicate that you can generate them from the actual code, and some editors auto-generate them. But how would you use them, as a replacement for the API documentation? To manually read them? Should they be commited to the `audbackend` repository, even though they can be generated on the fly?

Reply:
But I agree, that the typing is not so easy with classes. What this should indicate here is, that you have to provide one backend class, but doesn't matter which (of course it should not be `Base`, but one of the others that implement a backend):

![image](https://github.com/audeering/audb/assets/173624/9a014165-e333-4d06-97ce-14c099e3a586)

Reply:
> Interesting, I wasn't aware of `.pyi` files. I also have not completely understood them. https://mypy.readthedocs.io/en/stable/stubs.html seems to indicate that you can generate them from the actual code, and some editors auto-generate them. But how would you use them, as a replacement for the API documentation? To manually read them? Should they be commited to the `audbackend` repository, even though they can be generated on the fly?

This was just an idea as I saw the abstract types gaining in complexity quite a lot - and I  started to feel uncomfortable with this complexity.

However, reading up on the topic I found on [SOV](https://stackoverflow.com/questions/68079405/how-to-get-sphinx-to-use-type-annotations-from-python-stub-files) and [this + follow ups](https://github.com/sphinx-doc/sphinx/issues/7630#issuecomment-1378479217) that this does not got well together with `autodoc`.

So probably it will not be so beneficial to try to incorporate such a featrure.  In the C/C++ world there also is the distintion between .h[pp]  and c[pp] files. Often both of them are documented, with the header files often containing documentation from an api kind of perspective, i.e. being more "bird's eye".  I found that I needed quite some time to get a handle on the architecture by only reading the code - so I would find it good to extend the high level documentation of especially audbackend - possibly even augment it by (class)diagrams.  

So in short, I am taking my comment back - and will be closing the conversation here. Instead I will refer to this aspect in the general review.

Reply:
>>  of course it should not be Base, but one of the others that implement a backend

I wonder whether this would not be a prototypical case for an Abstract Base Class (that you don't want to be implemented). I mean on method level you can always raise a `NotImplementedError`, but without ABC you afaics cannot  prevent instantiation of a class.

Reply:
>  I found that I needed quite some time to get a handle on the architecture by only reading the code - so I would find it good to extend the high level documentation of especially audbackend - possibly even augment it by (class)diagrams.

I would start reading the compiled API documentation and the usage docs first. Unfortunately, you need to compile it yourself as `audbackend` 2.0.0 is not released yet.

Reply:
The concept in `audbackend` is that we now have two layers.

We have backend classes, that need to derive from `audbackend.backend.Base`, that are responsible for connecting to a backend, and transfering files between the backend and your PC, currently these involve:

* `audbackend.backend.Artifactory`
* `audbackend.backend.FileSystem`

As a developer, you can use them to directly put files to a backend, but usually this is not recommended.
The only methods a developer should call directly are `open()` and `close()` or use a `with` statement, to establish a connection to a backend.

The next higher abstraction level over the backends are backend interfaces, which you provide a backend object during initialization. These take care about the structure (or other stuff you could imagine might be useful to implement, like a user rights management layer) the data are actually stored on the backend.
All backend interfaces need to derive from `audbackend.interface.Base`. Currently we have:

* `audbackend.interface.Unversioned`
* `audbackend.interface.Versioned`
* `audbackend.interface.Maven`

The first one, just puts the files directly on the backend, as if a user would directly use the backend classes. The other two require a `version` argument to most of its methods, e.g. `put_file()`. It uses then the version and the file name to store the file at a unique location on the backend, that is usually hidden from the user.

Reply:
>> I would start reading the compiled API documentation and the usage docs first. Unfortunately, you need to compile it yourself as audbackend 2.0.0 is not released yet.

Yes I had come across this - they were not going together any longer. 

>> I think my approach to base classes in Python so far was, that I don't care about this. 

The have the benefit of being self-documenting and throwing an error when instantiated. I think for now it would be better to not pollute the issue backlog with this - and wait until we next time stumble across it. I myself know it now. 

About the section that starts with

>> The concept in audbackend is that we now have two layers.

This is nicely written. I wish I had had it when I started reading up on this :-) 

Still this imo is ok - resolving thread.